File: src/cjhx_agile_workflow/audit.py
from __future__ import annotations

from typing import Any


SENSITIVE_KEY_PARTS = (
    "authorization",
    "credential",
    "password",
    "privatekey",
    "secret",
    "token",
    "apikey",
    "api_key",
)


def redact(value: Any) -> Any:
    """Create a JSON-compatible audit copy with common credential fields removed."""
    if isinstance(value, dict):
        result: dict[str, Any] = {}
        for key, item in value.items():
            normalized = str(key).lower().replace("-", "").replace("_", "")
            sensitive = any(
                part.replace("_", "") in normalized for part in SENSITIVE_KEY_PARTS
            )
            result[str(key)] = "[REDACTED]" if sensitive else redact(item)
        return result
    if isinstance(value, list):
        return [redact(item) for item in value]
    if isinstance(value, tuple):
        return [redact(item) for item in value]
    return value

File: src/audit.ts
import type { JsonValue } from "./models.js";

const sensitive = ["authorization", "credential", "password", "privatekey", "secret", "token", "apikey"];

export function redact(value: JsonValue): JsonValue {
  if (Array.isArray(value)) return value.map(redact);
  if (typeof value === "object" && value !== null) {
    return Object.fromEntries(Object.entries(value).map(([key, item]) => {
      const normalized = key.toLowerCase().replaceAll(/[-_]/g, "");
      return [key, sensitive.some((part) => normalized.includes(part)) ? "[REDACTED]" : redact(item)];
    }));
  }
  return value;
}
